docs/dml-language-reference.md

```diff
@@ -1536,6 +1537,62 @@ Examples:
+features | Training features | matrix | yes 
+labels | Training labels | matrix | yes
+val_features | Validation features | matrix | yes
+val_labels | Validation labels | matrix | yes
+upd | Physical name of gradient calculation function. The format should be "related path:func name". For example, "./mnist_lenet_paramserv_sgd.dml::gradients". | string | yes
+agg | Physical name of gradient aggregation function. The format should be "related path:func name". For example, "./mnist_lenet_paramserv_sgd.dml::aggregation". | string | yes
+mode | Execution backend where the parameter function is executed | string | yes | "LOCAL", "REMOTE_SPARK"
+utype | Update strategy | string | yes | "BSP", "ASP"
+freq | Frequency of parameters updating. When setting to "BATCH", the gradients will be pushed to server for updating the parameters after each mini-batch. When setting to "EPOCH", the gradients will be pushed to server after each epoch. | string | no | "BATCH"(default), "EPOCH"
+epochs | Number of epoch | integer | yes |
+batchsize | Size of mini-batch, if the update frequence is setted to "EPOCH", this parameter will be ignored. | integer | no | 64(default)
+k | Degree of parallelism | integer | no | Number of vcores, otherwise vcores / 2 if using openblas(default)
+scheme | Scheme of data partition, i.e., how the data is distributed across workers | string | no | "DISJOINT_CONTIGUOUS", "DISJOINT_ROUND_ROBIN", "DISJOINT_RANDOM", "OVERLAP_RESHUFFLE"
+hyperparams | Additional hyper parameters, e.g., learning rate, momentum | list | yes | 
+checkpointing | Checkpoint strategy, currently not supported | string | no | 
+
+**Table**: Output of paramserv function
+
+Output | Description | Type
+-------- | ----------- | ----------
+model | Trained model | list
+
+**Update function:**
+
+The update function aims to apply the input training data to calculate the gradients for the given model. The implementation of this function should be based on a function signature like this: (i.e., **the input parameter including both type and name should be exactly the same as the below, except that the output name could be different**)
+
+```sh
+gradients = function(matrix[double] features, matrix[double] labels,
+                    list[unknown] hyperparams, list[unknown] model)
+          return (list[unknown] gradients)
+          # the output name can be something else than "gradients" but should always return a list
+```
+
+**Aggregate function:**
+
+The aggregate function aims to leverage some optimizers to update the model with the given gradients. The implementation of this function should be based on a function signature like this: (i.e., **the input parameter including both type and name should be exactly the same as the below, except that the output name could be different**)
+
+```sh
+aggregation = function(list[unknown] model, list[unknown] gradients,
+                       list[unknown] hyperparams)
```

Comment:
Could we please cleanup this API and use the same order for model and hyper parameters in both functions?